### DISCUSSION TITLE
CLI: variables in filenames appear incorrectly

### DISCUSSION THREAD

**Discussion by geldrin** [ImageMagick/ImageMagick#5579] [Help] [answered]
Recently I learned that IM is capable to put various input, output and image properties into the output filename, so I figured out I can save up an additional `identify` cycle in my workflow if I could put the desired parameters into the output filename and just parse that.

I want to use the output **resolution**, the **input filename**, and the **`fx:mean`** value in the output filename.

This is what I tried:

    convert tests/000005.jpg -precision 2 -set filename:m '%[fx:mean]' -resize 200% +dither -colors 2 -colorspace gray -contrast-stretch 0 -trim +repage -set filename:f '%[t]' -set filename:r '%[w]x%[h]' -border 5 -bordercolor '#000' ../tests/\%[filename:f]-\%[filename:m]-\%[filename:r].png
    
**Command breakdown:**

    convert tests/000005.jpg \       # input filename
      -precision 2 \                    # shorter presision for shorter mean value
      -set filename:m '%[fx:mean]' \    # mean value of input image
      -resize 200% +dither -colors 2 -colorspace gray -contrast-stretch 0 -trim +repage \  # image processing block
      -set filename:f '%[t]' \          # store basename of the input file ("000005")
      -set filename:r '%[w]x%[h]' \     # store output resolution in 'WIDTHxHEIGHT' format
      -border 5 -bordercolor '#000' \   # add border
      tests/%[filename:f]-%[filename:m]-\%[filename:r].png  # set output filename using the supplied variables

**Test system:** Debian 4.19.194-2
**Shell:** bash 5.0.3(1)-release (x86_64-pc-linux-gnu)
**Imagick version:** ImageMagick 6.9.10-23 Q16 x86_64 20190101

Expected output filename: `000005-0.086-2410x2017.png`
Output filename: `'000005-%0.086-%[2410x2017.png'`

As you can see, the variables are parsed, but the `%`, `[` and `]` characters are resolved in a seemingly random manner.

If I combine variables into a string like this: `-set filename:f '%[t]_%[fx:mean]_'`, the program also makes some characters disappear.
**Command:** 

    convert tests/000001.jpg -precision 2 -set filename:f '%[t]_%[fx:mean]_' -resize 200% +dither -colors 2 -colorspace gray -contrast-stretch 0 -trim +repage -set filename:r '%[w]x%[h]' -border 5 -bordercolor '#000' tests/%[filename:f]_%[filename:r].png

Expected filename: `000005_0.086_2410x2017.png`
Output filename: `000005_02410x2017.png`

In this example, the special characters disappeared (`'`, `%`, `[`), but some underscores were removed by the program.
I found this behavior really confusing and I have no idea what am I doing wrong... Did anyone had an issue with this?

**Comment by snibgo** [comment, marked as answer]
IM used to have problems with multiple "`filename:X`" in the output filename, but that was fixed some while ago.

Either method works fine in recent 7 IM:
```
magick -precision 2 toes.png -set filename:f %[t] -set filename:m %[fx:mean] -set filename:r %[w]x%[h] abc-%[filename:f]-%[filename:m]-%[filename:r].jpg

magick -precision 2 toes.png -set filename:f %[t]-%[fx:mean]-%[w]x%[h] abc-%[filename:f].jpg
```
In both cases, the created filename is `abc-toes-0.49-267x233.jpg`. (Tested on IM v7.1.0-47 o Windows 8.1.)

If you can upgrade to v7, I suggest doing so.

If your v6 IM isn't the latest v6, then I suggest upgrading to the latest v6.

**Reply by geldrin** [reply]
Thanks for the answer, I also had the impression that the issue was caused by a bug.

Fortunately, **_I found a workaround_** by combining the parameters into a single variable, then use that as a single `filename` tag at the end (which proves your claim, that the issue is caused by multiple filename tags).

So this command works with IM 6.9.x:

    convert 001.jpg -quiet -set filename:m '%[fx:mean]' -resize 200% +dither -colors 2 -colorspace gray -contrast-stretch 0 -trim +repage -set filename:f '%[t]_%[filename:m]_%[w]x%[h]' -border 5 -bordercolor '#000' "%[filename:f].png"

**Comment by GeeMack** [comment]
Just adding a comment about a feature of `-set filename:X some_name` that I find useful. A single name like `filename:f` can be set on each of several output images. Each image set to that `filename:f` variable can be set to its own separate file name. Then when `%[filename:f].png` is written at the end of the command, it will output all the images each named with what you set above.

    convert rose: -set filename:f size_%[w]_original \
       \( -clone 0 -resize 200% -set filename:f size_%[w]_medium \) \
       \( -clone 0 -resize 300% -set filename:f size_%[w]_large \) \
       %[filename:f].png

That command will output three images, `size_70_original.png`, `size_140_medium.png`, and `size_210_big.png`.
